### PR TITLE
Recipes for optimzing training scripts

### PR DESCRIPTION
**Pull request by LoserCheems**
R1 series is very popular, many new R1 models are coming, we need a `recipes folder` to put different model configurations to optimize the training command line.

### CODE REVIEW

File: recipes/qwen/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+> [!NOTE]
+> The training commands below are configured for a node of 8 x H100s (80GB). For different hardware and topologies, you may need to tune the batch size and number of gradient accumulation steps.
+
+```shell
+# SFT
+ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py recipes/qwen/sft/config_qwen2.5-1.5B.yaml
+
+# GRPO
+ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml scripts/grpo.py recipes/qwen/grpo/config_qwen-7B.yaml
```
Comment (edbeeching) [inline recipes/qwen/README.md:21]:
```suggestion
ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_rl/grpo.py recipes/qwen/grpo/config_qwen-7B.yaml
```

File: recipes/qwen/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+## Training
+
+We support training models with either DDP or DeepSpeed ZeRO-2 and ZeRO-3. To switch between methods, simply change the path to the `recipes` YAML config in `accelerate_configs`.
+
+> [!NOTE]
+> The training commands below are configured for a node of 8 x H100s (80GB). For different hardware and topologies, you may need to tune the batch size and number of gradient accumulation steps.
+
+```shell
+# SFT
+ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py recipes/qwen/sft/config_qwen2.5-1.5B.yaml
```
Comment (edbeeching) [inline recipes/qwen/README.md:18]:
```suggestion
ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py --config recipes/qwen/sft/config_qwen2.5-1.5B.yaml
```

File: src/open_r1/sft.py [hunk 3]
```diff
@@ -50,34 +60,81 @@
...
+    # Load datasets
+    ################
+    dataset = load_dataset(script_args.dataset_name, name=script_args.dataset_config)
+
     ################
-    # Model init kwargs & Tokenizer
+    # Load tokenizer
     ################
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_args.tokenizer_name_or_path, trust_remote_code=model_args.trust_remote_code, use_fast=True
```
Comment (edbeeching) [inline src/open_r1/sft.py:111]:
```suggestion
        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, use_fast=True
```

File: recipes/qwen/sft/config_qwen2.5-1.5B.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+max_steps: -1
+num_train_epochs: 1
+output_dir: data/Qwen2.5-1.5B-Open-R1-Distill
+overwrite_output_dir: true
+per_device_eval_batch_size: 4
+per_device_train_batch_size: 4
+push_to_hub: true
+remove_unused_columns: true
+report_to:
+- tensorboard
```
Comment (edbeeching) [inline recipes/qwen/sft/config_qwen2.5-1.5B.yaml:42]:
```suggestion
```

File: recipes/qwen/grpo/config_qwen-7B.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+max_steps: -1
+num_train_epochs: 1
+output_dir: data/DeepSeek-R1-Distill-Qwen-7B-GRPO
+overwrite_output_dir: true
+per_device_eval_batch_size: 4
+per_device_train_batch_size: 1
+push_to_hub: true
+remove_unused_columns: true
+report_to:
+- tensorboard
```
Comment (edbeeching) [inline recipes/qwen/grpo/config_qwen-7B.yaml:41]:
```suggestion
```

File: recipes/qwen/grpo/config_qwen-7B.yaml
Comment (lewtun) [inline recipes/qwen/grpo/config_qwen-7B.yaml:1]: Given all the variants of Qwen and Llama, let's rename the recipe path to include the model ID in the name like this:

```
recipes/DeepSeek-R1-Distill-Qwen-7B/grpo/config_full.yaml
```

the `full` refers to full training so that others can contribute e.g. `config_lora.yaml` at a later point

File: recipes/qwen/sft/config_qwen2.5-1.5B.yaml
Comment (lewtun) [inline recipes/qwen/sft/config_qwen2.5-1.5B.yaml:1]: Similar to the above, let's use the model ID to indicate which model we're optimising, i.e.

```
recipes/Qwen/Qwen2.5-1.5B-Instruct/sft/config_full.yaml
```

**Review by lewtun — CHANGES_REQUESTED** [review]
Thanks a lot for the very clean PR @LoserCheems ! Overall it looks great, with just some nits on my side about using the model ID to indicate the recipe path more clearly.

**Comment by kashif** [comment]
@LoserCheems thanks, My apologies for the grpo merge conflicts... could you kindly have a look at the new configuration arguments, `--use_vlllm` as well as the `num_processes`

**Comment by LoserCheems** [comment]
Thank you @edbeeching @lewtun @kashif, I am working on these issues and will update soon!

File: recipes/qwen/grpo/config_qwen-7B.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+lr_scheduler_type: cosine
+max_prompt_length: 256
+max_steps: -1
+num_train_epochs: 1
+output_dir: data/DeepSeek-R1-Distill-Qwen-7B-GRPO
+overwrite_output_dir: true
+per_device_eval_batch_size: 4
+per_device_train_batch_size: 1
+push_to_hub: true
+remove_unused_columns: true
```
Comment (edbeeching) [inline recipes/qwen/grpo/config_qwen-7B.yaml:39]:
```suggestion
```
We need to keep the `solutions` column as it is used by the reward fn.

**Review by edbeeching — CHANGES_REQUESTED** [review]
Thanks for adding this, a few things were broken so I have suggested some changes. 

I can confirm that both the SFT and GRPO scripts run with the changes I proposed.

Please can you also rebase on main as there were some eval callbacks added to the SFT and GRPO training scripts

**Comment by kashif** [comment]
@LoserCheems reopening and then i can also help in the merge conflicts

File: README.md [hunk 7]
```diff
@@ -114,30 +99,13 @@ Here `{model}` and `{dataset}` refer to the model and dataset IDs on the Hugging
 
 ### GRPO
 
-To train via the GRPO trainer we will use the strategy of using one node to run vLLM for faster generation and the remaining nodes for training. Thus we will use the `configs/zero3.yaml` config and then overwrite the `num_processes=7` for the 8 GPU training scenario. Thus all we need to do is:
+To train via the GRPO trainer we will use the strategy of using one node to run vLLM for faster generation and the remaining nodes for training. Thus we will use the `configs/zero3.yaml` config and then overwrite the `num_processes=8` for the 8 GPU training scenario. Thus all we need to do is:
```
Comment (kashif) [inline README.md:102]:
```suggestion
To train via the GRPO trainer we will use the strategy of using one node to run vLLM for faster generation and the remaining nodes for training. Thus we will use the `configs/zero3.yaml` config and then overwrite the `num_processes` to `num_processes=7` for the 8 GPU training scenario. Thus all we need to do is:
```

**Comment by LoserCheems** [comment]
Thank you @kashif,  I have successfully dealt with these issues.

File: README.md [hunk 8]
```diff
@@ -114,30 +99,13 @@ Here `{model}` and `{dataset}` refer to the model and dataset IDs on the Hugging
...
-    --use_vllm \
-    --vllm_device auto \
-    --vllm_gpu_memory_utilization 0.7
-```
-
-To launch a Slurm job, run:
-
-```shell
-sbatch --output=/path/to/logs/%x-%j.out --err=/path/to/logs/%x-%j.err slurm/grpo.slurm {model} {dataset} {accelerator}
+ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/grpo.py --config recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml
```
Comment (kashif) [inline README.md:105]:
```suggestion
ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/grpo.py --config recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml
```
To launch a Slurm job, run:

```shell
sbatch --output=/path/to/logs/%x-%j.out --err=/path/to/logs/%x-%j.err slurm/grpo.slurm {model} {dataset} {accelerator}
```

**Comment by LoserCheems** [comment]
Sorry @edbeeching, I've removed the model recipe.

File: README.md [hunk 9]
```diff
@@ -128,22 +113,10 @@ Here `{model}` and `{dataset}` refer to the model and dataset IDs on the Hugging
 
 ### GRPO
 
-To train via the GRPO trainer we will use the strategy of using one node to run vLLM for faster generation and the remaining nodes for training. Thus we will use the `configs/zero3.yaml` config and then overwrite the `num_processes=7` for the 8 GPU training scenario. Thus all we need to do is:
+To train via the GRPO trainer we will use the strategy of using one node to run vLLM for faster generation and the remaining nodes for training. Thus we will use the `configs/zero3.yaml` config and then overwrite the `num_processes` to `num_processes=7` for the 8 GPU training scenario. Thus all we need to do is:
```
Comment (lewtun) [inline README.md:116]: nit: 
"node" has a special meaning (= N GPUs), so let's use this:
```suggestion
To train via the GRPO trainer, we use one GPU to run vLLM for faster generation and the remaining GPUs for training. For example, one a node with 8 GPUs, use the `recipes/accelerate_configs/zero3.yaml` config and then overwrite `num_processes` to run on 7 devices:
```

File: README.md [hunk 10]
```diff
@@ -128,22 +113,10 @@ Here `{model}` and `{dataset}` refer to the model and dataset IDs on the Hugging
...
-    --max_prompt_length 512 \
-    --max_completion_length 1024 \
-    --per_device_train_batch_size 1 \
-    --gradient_accumulation_steps 16 \
-    --logging_steps 10 \
-    --bf16 \
-    --use_vllm \
-    --vllm_device auto \
-    --vllm_gpu_memory_utilization 0.7
+ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/grpo.py --config recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml
```
Comment (lewtun) [inline README.md:119]:
```suggestion
ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/zero3.yaml --num_processes=7 src/open_r1/grpo.py --config recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml
```

**Review by lewtun — COMMENTED** [review]
Thanks for iterating @LoserCheems ! The changes look good and I've left some minor nits. I'm just checking all commands / configs run on our side and if that works then I think we can merge

File: README.md (hunk 10 above)
Comment (kashif) [reply README.md]: so the grpo recipes already have the num_processes set to `7`

Comment (LoserCheems) [reply README.md]: should we reverse this change?

File: recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml [hunk 11]
```diff
@@ -0,0 +1,45 @@
+# Model arguments
+model_name_or_path: <your-hub-id>/Qwen2.5-1.5B-Open-R1-Distill
```
Comment (lewtun) [inline recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml:2]: nit to ensure the example in the README runs:
```suggestion
model_name_or_path: deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B
```

**Review by lewtun — APPROVED** [review]
OK I found one small bug, but after we fix that this is good to go! 

Thank you so much for this very clean contribution @LoserCheems 🚀

File: recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml (hunk 11 above)
Comment (LoserCheems) [reply recipes/qwen/Qwen2.5-1.5B-Instruct/grpo/confg_full.yaml:2]: `DeepSeek-R1-Distill-Qwen-1.5B` uses `Qwen2.5-Math-1.5B` as the base model, while the current sft uses `Qwen2.5-1.5B-Instruct`. I'm not sure if that leads to some bias.

**Review by kashif — APPROVED** [review]
